### PR TITLE
[inductor] structured-log graph execution order + test

### PR DESCRIPTION
**Pull request by skarjala**
Summary:

- Emit a structured trace per compiled graph execution to reconstruct execution order in TLParse.
- Adds debug.log_graph_execution(name) called from `CompiledFxGraph.__call__`, producing an artifact named inductor_graph_execution with payload {"graph": "graph_<id>"}.

Testing:
- Add inline test to verify structure and output

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160448

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

### CODE REVIEW

File: test/dynamo/test_structured_trace.py [hunk 1]
```diff
@@ -1475,6 +1475,51 @@ def f(x):
...
+    def test_graph_execution_simple(self):
+        class SimpleModule(torch.nn.Module):
+            def forward(self, x):
+                return torch.relu(x)
+
+        with self._setup_graph_execution_capture() as payload_buffer:
+            torch._dynamo.reset()
+            mod = SimpleModule()
+            compiled = torch.compile(mod, backend="inductor")
+            compiled(torch.randn(2, 2))
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: try calling it multiple times, you should be seeing 1 entry per call

File: torch/_inductor/debug.py [hunk 2]
```diff
@@ -788,6 +788,25 @@ def dtype_to_str(dtype: Any) -> Optional[str]:
...
 
 
+def log_graph_execution(name: str) -> None:
+    """Log a single graph execution for TLParse."""
+
+    try:
+        trace_structured(
+            "artifact",
+            metadata_fn=lambda: {
+                "name": "inductor_graph_execution",
```
Comment (xmfan) [inline torch/_inductor/debug.py]: it's not a good idea to generate 1 artifact per graph call. jobs can have hundreds of graphs, each called thousands of times

Comment (skarjala) [reply torch/_inductor/debug.py]: fixed. now i first accumulate in a list called _graph_execution_order in every graph call, and only dump it once at the end.

File: test/dynamo/test_structured_trace.py [hunk 3]
```diff
@@ -1501,20 +1501,24 @@ def forward(self, x):
...
-            mod = SimpleModule()
-            compiled = torch.compile(mod, backend="inductor")
-            compiled(torch.randn(2, 2))
+            mod1 = SimpleModule()
+            compiled1 = torch.compile(mod1, backend="inductor")
+            compiled1(torch.randn(2, 2))
+            mod2 = SimpleModule()
+            compiled2 = torch.compile(mod2, backend="inductor")
+            compiled2(torch.randn(2, 2))
+            torch._inductor.debug.log_graph_execution()
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: it isn't really a test if you manually call the function the compiled code is supposed to call

File: torch/_inductor/debug.py [hunk 4]
```diff
@@ -802,23 +803,25 @@ def dtype_to_str(dtype: Any) -> Optional[str]:
...
-        log.debug(
-            "Failed to log inductor_graph_execution via structured logging",
-            exc_info=True,
-        )
+        log.debug("Failed to log inductor_graph_execution", exc_info=True)
+
+
+# Dump accumulated graph executions on process exit
+if config.log_tlparse:
+    atexit.register(log_graph_execution)
```
Comment (xmfan) [inline torch/_inductor/debug.py:824]: i don't really understand what's going on here. why aren't we adding the log into the inductor output code?

File: test/dynamo/test_structured_trace.py [hunk 5]
```diff
@@ -1501,20 +1501,24 @@ def forward(self, x):
...
 
-            # Assert payload inline with dynamic graph name normalized
             payload_content = payload_buffer.getvalue().strip()
-            normalized = re.sub(
-                r"(\"graph\":\s*)\"[^\"]+\"", r'\1"GRAPH"', payload_content
-            )
+            normalized = re.sub(r"graph_\d+", "GRAPH", payload_content)
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: why are we normalizing? is the graph_id not resetting between tests?

File: torch/_inductor/debug.py (hunk 4 above)
Comment (xmfan) [inline torch/_inductor/debug.py]: To simplify things, how about we provide a user facing api to record their graph execution order? For instance,

```python
with torch._inductor.debug.record_and_log_graph_execution_order():
  for i in range(10):
    compiled1()
    compiled2()
```

and on context manager exit, you can call log_graph_execution

File: torch/_inductor/output_code.py [hunk 6]
```diff
@@ -55,6 +55,9 @@
 from .runtime.autotune_cache import AutotuneCacheBundler
 
 
+_graph_execution_order: list[str] = []
+_record_graph_execution: bool = False
+
```
Comment (xmfan) [inline torch/_inductor/output_code.py]: could we move these into debug.py

File: torch/_inductor/output_code.py [hunk 7]
```diff
@@ -581,6 +584,11 @@ def __del__(self) -> None:
 
     def __call__(self, inputs: Sequence[Any]) -> Any:
         assert self.current_callable is not None
+
+        graph_id = self.fx_kwargs.get("graph_id")
+        name = f"graph_{graph_id}" if graph_id is not None else "unknown"
```
Comment (xmfan) [inline torch/_inductor/output_code.py]: how does graph_id map to compile id?

(do you have graph_id in tlparse?)

File: torch/_inductor/compile_fx.py [hunk 8]
```diff
@@ -1619,6 +1619,7 @@ def codegen_and_compile(
                         static_input_idxs,
                         graph_kwargs,
                         inputs_to_check,
+                        str(torch._guards.CompileContext.current_compile_id()),
```
Comment (xmfan) [inline torch/_inductor/compile_fx.py]: @oulgen i think this might prevent cache hits? it's a bit strange that CI is green

Comment (oulgen) [reply torch/_inductor/compile_fx.py]: Outputgraph is the result of the cache, so i think it is fine since it does not affect cache key, but when this is cached it will just be wrong since compile id can change

Comment (xmfan) [reply torch/_inductor/compile_fx.py]: Synced offline, should be fine as long as user's program is single threaded, and we can handle cache case by adding similar logic to codecache.py

File: test/dynamo/test_structured_trace.py [hunk 9]
```diff
@@ -1489,6 +1489,47 @@ def f(x):
...
+
+            with record_and_log_graph_execution_order():
+                compiled(torch.randn(1))
+
+            payload_content = payload_buffer.getvalue().strip()
+            payload = json.loads(payload_content)
+            executions = payload["graph_execution_order"]
+            self.assertTrue(all(isinstance(e["compile_id"], str) for e in executions))
+            self.assertExpectedInline(
+                json.dumps(payload, sort_keys=True),
```
Comment (xmfan) [inline test/dynamo/test_structured_trace.py]: sort_keys? why is that needed. it defeats the purpose of this test if we can't log compile ids in order....

**Review by xmfan — APPROVED** [review]
